### ISSUE TITLE
Fonts vertical position shifted after 3.4.0

### ISSUE THREAD

**Issue by ghost** [adventuregamestudio/ags#888]
Easily seen in, for example, "Harry Potter RPG": https://www.adventuregamestudio.co.uk/site/games/game/2358/
https://drive.google.com/file/d/18_8Y7zs-cg47Wn10NX641vCjV-Rv3y2A/view
**EDIT:** the link is broken, but I found slightly older version of the game, which still has same problem:
https://archive.org/details/harry-potter-rpg

On the character creation screen all button texts are supposed to be centered, but instead they are lowered down by noticeable amount.

Test shows that this regression occured after commit be5750094d5b4838ddd586289f4e7b25cf965522 (early 3.4.1 build) which removed prebuilt alfont.lib from Windows version and since when engine depends on alfont library built from our repository (https://github.com/adventuregamestudio/lib-alfont).

This error persists in latest 3.5.0 which has corresponding alfont + freetype sources integrated.

**Comment by rofl0r** [comment]
could it be due to the different versions of alfont used by engine/libsrc vs libalfont repo ?
iirc the latter one was used for the windows build.

**Comment by ivan-mogilko** [comment]
To clarify, there were 3 stages:
1) prebuilt alfont + freetype made by Chris Jones which Windows version of AGS used before 3.4.1.
2) our own fork of alfont + freetype (https://github.com/adventuregamestudio/lib-alfont) to which we switched at 3,4.1
3) current alfont + freetype sources in the main repo (since 3.5.0).

The bug occurs in 2) and 3), which could mean that even our alfont fork was already somewhat different from original prebuilt binary by CJ.

**Comment by fernewelten** [comment]
There are several heights associated with fonts, the baseline height being the most famous. The alfont library might have started out simply stringing letters together as if they were boxes, and when it matured and integrated more of proper typography, migrated from letter height to baseline height. That would explain why all the letters moved down a little bit.

**Comment by ivan-mogilko** [comment]
Yes, iirc there was a hack by CJ that discarded descend height from the fonts, cutting their height.
https://github.com/adventuregamestudio/lib-alfont/commit/089f0e59c8a200503176dc355c5000c8fe9c6d20

**Comment by fernewelten** [comment]
So one way of looking at this would be that we have already changed the standard in version 3.4.1 in an incompatible way, albeit unwittingly. If we return to the world as it was in C.J.'s time, we'll break all the games that have been developed since version 3.4.1 up to now. If we continue with the new standard that has been set in 3.4.1, we'll break all the games that have been developed in olden times. 

If we want to make them **both** run on contemporary engines, one of the two groups would need to be fiddled with. For instance, we might perhaps extend the font definitions in our current engine with a “vertical offset” parameter. Flanging an old game onto a contemporary engine would entail tweaking these offsets to make the texts appear in just the right spot.

**Comment by fernewelten** [comment]
> a hack by CJ that discarded descend segment from the fonts, cutting their height.

Yes, that would **exactly** be the difference between baseline height and total height.

**Comment by rofl0r** [comment]
i think that the issue is this https://github.com/adventuregamestudio/lib-alfont/issues/3
the precompiled lib shipped for windows used the libalfont repo, whereas the engine sources used a different version (which was only used on linux and mac until then).

**Comment by morganwillcock** [comment]
So currently, what is described in #447 is accidentally fixed right now?

**Comment by ivan-mogilko** [comment]
>  For instance, we might perhaps extend the font definitions in our current engine with a “vertical offset” parameter.  Flanging an old game onto a contemporary engine would entail tweaking these offsets to make the texts appear in just the right spot.

This parameter exists, and was originally added for named purpose.

**Comment by ivan-mogilko** [comment]
Hmm... don't know how did I miss this earlier, but maybe the license file distributed with AGS hints what was the original version of Alfont and FreeType libs used by AGS prior to 3.4.1:

https://github.com/adventuregamestudio/ags/blob/release-3.4.0/Windows/Installer/Source/Docs/license_alfont.txt

Note it is saying Alfont 1.5 and lower `I am NOT responsable of the contents of the freetype-2.0.5`.

**Comment by rofl0r** [comment]
source can be found here. it should be easy to create a recursive diff (`diff -ru old-dir new-dir`) vs our freetype sources, remove all non-.c/.h hunks and then apply it on top our embedded copy, to build and test whether it fixes the issue.
https://sourceforge.net/projects/gnuwin32/files/freetype/2.0.5/freetype-2.0.5-src.zip/download

**Comment by ivan-mogilko** [comment]
I was experimenting with alfont lately, trying to solve a separate issue, and this regression may be fixed by restoring the original commented out line that initialized `face_ascender` value here:
https://github.com/adventuregamestudio/ags/blob/69944487a2ff3cab9fb40f8c0d31bbf986dab1a1/Common/libsrc/alfont-2.0.9/alfont.c#L470-L475

This results in a perfect look in this game matching original windows executable (from AGS 3.3.4).

It's possible that this also resolves #447, but unfortunately there's no real game attached for testing, so I cannot be certain.

These hacks we have in alfont source are attempts to reverse engineer a precompiled alfont lib used by an older windows engine. Maybe they were not accurate, and there's also a possibility that alfont library was modified between 2.72 and 3.* versions as well, leading to further incompatibilities.

I will run few tests over older games to see if anything will broke if I restore this line.

---

On a related note, to think of it, I have absolutely no idea why these changes are at all necessary. They were made by @jjsat years ago, but any information regarding this is probably in his old mantis bug tracker circa 2011-12.

Maybe we could try restore the original code and see if it breaks any old games,

**Comment by ivan-mogilko** [comment]
So, if I restore the `ascender` field calculation, then there's an opposite problem in another game "The Order of the Thorne" (demo could be found on steam, and maybe somewhere else).

The most curious is, though, that both games are made *with the same version of AGS* (3.3.4.2), and original Windows executables displayed everything correctly.

**Comment by morganwillcock** [comment]
> It's possible that this also resolves #447 reported by @morganwillcock , but unfortunately there's no real game attached for testing, so I cannot be certain.

If it helps to test anything it was this one: https://gamejolt.com/games/theburgomaster/344123

**Comment by ivan-mogilko** [comment]
@morganwillcock I assume the game already has some fixes to font position? Because I don't see anything being cut of from below. But looks like it on contrary is a bit cut of from the top.

**Comment by morganwillcock** [comment]
I think it needs the high resolution to show how large the vertical shift is. It was built with these font settings:
```
      <Font>
        <ID>0</ID>
        <PointSize>36</PointSize>
        <Name>fntDefault</Name>
        <OutlineFont>0</OutlineFont>
        <OutlineStyle>Automatic</OutlineStyle>
        <SourceFilename>C:\Users\mwillcock\Downloads\Berylium\Berylium.ttf</SourceFilename>
        <VerticalOffset>-2</VerticalOffset>
        <LineSpacing>0</LineSpacing>
      </Font>
      <Font>
        <ID>1</ID>
        <PointSize>36</PointSize>
        <Name>fntSpeech</Name>
        <OutlineFont>0</OutlineFont>
        <OutlineStyle>Automatic</OutlineStyle>
        <SourceFilename>C:\Users\mwillcock\Downloads\vollkorn\Vollkorn-Regular.ttf</SourceFilename>
        <VerticalOffset>-2</VerticalOffset>
        <LineSpacing>0</LineSpacing>
      </Font>
```
The -2 was to stop the clipping in-game.

**Comment by ivan-mogilko** [comment]
Alright... so I've been considering this:
* alfont code without hacks draws fonts correctly in game A but not game B;
* alfont code with hacks draws fonts incorrectly in game A, but suddenly makes them correct again in B.
* older Windows engine was drawing them all correctly (or at least as we expect to keep compatibility).

This makes apparent that there is some factor that makes a font require this hack. Other fonts should not be applied a hack.

So, I started checking font parameters set by FreeType under debugger, and noticed that the fonts that got broken by a correct alfont code *had its real pixel height smaller than requested*. While the fonts that were drawn correctly got the real height equal or larger than requested.

I realize that this is a completely random assumption, but decided to make a test and did this:
<pre>
int real_height = abs(f->face->size->metrics.ascender >> 6) + abs(f->face->size->metrics.descender >> 6);

// Original alfont code
f->face_h = h; // where 'h' is a requested pixel height
f->real_face_h = test_h;
f->face_ascender = f->face->size->metrics.ascender >> 6;

// Our hack
if (real_height < h) {
    f->face_ascender = h;
}
</pre>

And now the text is drawn correctly in both "Harry Potter RPG" and "Order of the Thorn"!

**Comment by ivan-mogilko** [comment]
So I went ahead and pushed the fix: 41631638eca0bfab4306c631f5d0d0780889c3ad

Like said, not 100% sure that it's correct, but the more I think about it the more it seem to make sense.

Tested several games, old and newer, with different TTF fonts, and this commit produces expected results everywhere so far.

**Comment by ivan-mogilko** [comment]
Unfortunately the above fix appeared to be not enough. I missed this when testing earlier, but this is best seen using "MI 9-verb" or "Tumbleweed" templates:
- the speech font's outline are drawn with a slight vertical shift;
- in "Tumbleweed" fonts on the game options buttons appears also shifted up by 1-2 pixels, and no longer centered.

After noticing that I did more scrupulous test of other games I tried before (like "Order of the Thorne"), and found that they also have slight differences in font vertical position after my above changes.

This is annoying, because "Harry Potter RPG" seem the only one game which I found so far, where the compatibility hacks in alfont do *not* have to be applied. So its fonts might have to have some difference.

I went back to experimenting, and the only certain difference I noticed so far is the sizes of the fonts' "descender": in HP rpg the fonts have rather big descender parts for some reason.

I might use a dumb "descender < N" condition in the code based on these experiments, if won't find anything more logical...

**Comment by rofl0r** [comment]
> This is annoying, because "Harry Potter RPG" seem the only one game which I found so far, where the compatibility hacks in alfont do not have to be applied.

is it using a fixed version of the lucasarts font ? if so we might just add a hack that if the lucasarts font is detected, it is binary-patched to the fixed version.

**Comment by ivan-mogilko** [comment]
> is it using a fixed version of the lucasarts font ? if so we might just add a hack that if the lucasarts font is detected, it is binary-patched to the fixed version.

No, it's using a number of completely different fonts.

**Comment by ivan-mogilko** [comment]
Another option, I exported game fonts and the ones affected all have same certain family name. So, what I could do is to add some kind of a flag to alfont's function, which tells whether to apply the hack or not. For this particular font the engine will request to not apply the hack.

Other than that, i got no idea why it behaves differently. I double checked that the game was made using official engine, it's made with Windows version which still used original precompiled alfont library to which we did not get sources from Chris Jones. So either it had some additional hacks we do not know, or different version of FreeType.

**Comment by ivan-mogilko** [comment]
Alright, I found something else extremely weird.

I extracted the "HP" game fonts and made a test game with it, using same old version of the editor (3.3.4). Tried to replicate a random text from the real game, with same sizes etc. And at first it appeared that the test game made in original editor displays these fonts also "incorrectly", that is differently from the "HP", offsetting them vertically by a noteable amount.

But then I noticed that the "HP" has TTF anti-aliasing turned on, and after turning it on in the test game, the texts suddenly "jumped" up and appeared on the same places where they were in "HP".

To summarize, it seems now that the old windows engine (with prebuilt alfont) was rendering anti-aliased fonts at a different vertical offset compared to non-antialiased. This is something completely new to me.

**Comment by ivan-mogilko** [comment]
For the reference, here's the comparison of font output done by an original Windows 3.3.4.2 engine, and any post-3.4.1 engine (3.4.1 was the version where we replaced prebuilt alfont with one built from source in repo).

### Non-antialiased

3.3.4.2 on Windows:
![fonts-normal-3342](https://user-images.githubusercontent.com/1833754/140326150-b1cd9c3c-160f-4afb-83d8-f1e17825cc0e.png)
post-3.4.1:
![fonts-normal-post341](https://user-images.githubusercontent.com/1833754/140326204-877c1469-5368-4975-bf7f-5ae33b258ada.png)


### Anti-aliased

3.3.4.2 on Windows:
![fonts-antialiased-3342](https://user-images.githubusercontent.com/1833754/140326304-ef1c9684-7b3d-44cd-b77e-2c13b659495f.png)
post-3.4.1:
![fonts-antialiased-post341](https://user-images.githubusercontent.com/1833754/140326366-b81009ab-71c4-4716-afd7-e83dd43ef46a.png)

**Comment by ivan-mogilko** [comment]
After few more tests, I confirm: the original windows engine rendered anti-aliased TTF texts as if this "ascender hack" was not applied. Furthermore, as it is possible to switch font anti-aliasing at runtime (e.g. `SetGameOption(OPT_ANTIALIASFONTS, 0);`), turning it on and off will change the way fonts are offset vertically.

I do not see anything in the original engine code that would change fonts upon this option is set, so my only guess so far is that the prebuilt alfont lib contained additional changes that resulting in this effect.

I think this may be emulated by reupdating the TTF metrics when the AA option is assigned. Since we switched to open sourced alfont in 3.4.1, this extra hack would be required only for the games made with AGS <= 3.4.0.

**Comment by ivan-mogilko** [comment]
Okay, I think this is done now (14679a430bf29acb6e511bb3799a5f99d9929463).
